include/lbann/layers/regularizers/batch_normalization.hpp

```diff
@@ -49,6 +49,25 @@ enum class batch_normalization_stats_aggregation
   global
 };
 
+/**
+ * Define the accumulation type for batchnorm.
+ *
+ * This is the same as the input type, except for half, where it is
+ * float to ensure numerical stability.
+ *
+ * This is also type all weights are expected to be in.
+ */
+template <typename TensorDataType>
+struct BNAccT {
+  using type = TensorDataType;
+};
+#ifdef LBANN_HAS_GPU_FP16
+template <>
+struct BNAccT<fp16> {
+  using type = float;
+};
+#endif
```

Comment:
I recommend also adding a typedef to force the `::type`, which will save you some characters later on:
```
template <typename T>
using BNAcc = typename BNAccT<T>::type;
```
Now you can use `BNAcc<T>` anywhere you've used `typename BNAccT<T>::type` downstream (I notice a lot in the `.cu` file, e.g.).